allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,233 @@
+        span_end_probs = torch.nn.functional.softmax(span_end_logits, dim=-1)
+        best_spans = get_best_span(span_start_logits, span_end_logits)
+        best_span_scores = torch.gather(
+            span_start_logits, 1, best_spans[:, 0].unsqueeze(1)
+        ) + torch.gather(span_end_logits, 1, best_spans[:, 1].unsqueeze(1))
+        best_span_scores = best_span_scores.squeeze(1)
+
+        output_dict = {
+            "span_start_logits": span_start_logits,
+            "span_start_probs": span_start_probs,
+            "span_end_logits": span_end_logits,
+            "span_end_probs": span_end_probs,
+            "best_span": best_spans,
+            "best_span_scores": best_span_scores,
+        }
+
+        # Compute the loss for training.
+        if answer_span is not None:
+            span_start = answer_span[:, 0]
+            span_end = answer_span[:, 1]
+            span_mask = span_start != -1
+            self._span_accuracy(
+                best_spans, answer_span, span_mask.unsqueeze(-1).expand_as(best_spans)
+            )
+
+            start_loss = cross_entropy(span_start_logits, span_start, ignore_index=-1)
+            if torch.any(start_loss > 1e9):
+                logger.critical("Start loss too high (%r)", start_loss)
+                logger.critical("span_start_logits: %r", span_start_logits)
+                logger.critical("span_start: %r", span_start)
+                assert False
+
+            end_loss = cross_entropy(span_end_logits, span_end, ignore_index=-1)
+            if torch.any(end_loss > 1e9):
+                logger.critical("Start loss too high (%r)", end_loss)
+                logger.critical("span_end_logits: %r", span_end_logits)
+                logger.critical("span_end: %r", span_end)
+                assert False
+
+            loss = (start_loss + end_loss) / 2
+
+            self._span_start_accuracy(span_start_logits, span_start, span_mask)
+            self._span_end_accuracy(span_end_logits, span_end, span_mask)
+
+            output_dict["loss"] = loss
+
+        # Compute the EM and F1 on SQuAD and add the tokenized input to the output.
+        if metadata is not None:
+            best_spans = best_spans.detach().cpu().numpy()
+
+            output_dict["best_span_str"] = []
+            context_tokens = []
+            for metadata_entry, best_span in zip(metadata, best_spans):
+                context_tokens_for_question = metadata_entry["context_tokens"]
+                context_tokens.append(context_tokens_for_question)
+
+                best_span -= 1 + len(metadata_entry["question_tokens"]) + 2
+                assert np.all(best_span >= 0)
+
+                predicted_start, predicted_end = tuple(best_span)
+                assert predicted_start >= 0
+                assert predicted_end < len(context_tokens_for_question)
+
+                while (
+                    predicted_start >= 0
+                    and context_tokens_for_question[predicted_start].idx is None
```

Comment:
When does this happen?  If you selected a [CLS] or [SEP] token?

Reply:
When we couldn't figure out the token offsets for the token that was selected, for example, in the case of `"/ˈʃvɑːrtsnɛɡər/;[1][a] German: [ˈaʁnɔlt ˈʃvaʁtsn̩ˌʔɛɡɐ]"`.

Reply:
Since we don't allow people to pick their own tokenizer anymore, I shortened the warning in 213a4e14d93c8aa61eed77b266025e3bc8fe8141.